src/xrpld/app/tx/detail/Credentials.cpp

```diff
@@ -0,0 +1,471 @@
+    auto const subject = sle->getAccountID(sfSubject);
+    bool const accepted = sle->getFlags() & lsfAccepted;
+
+    auto err = delSLE(issuer, sfIssuerNode, !accepted || (subject == issuer));
+    if (!isTesSuccess(err))
+        return err;
+
+    if (subject != issuer)
+    {
+        err = delSLE(subject, sfSubjectNode, accepted);
+        if (!isTesSuccess(err))
+            return err;
+    }
+
+    // Remove object from ledger
+    view.erase(sle);
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialDelete::doApply()
+{
+    auto const subject = ctx_.tx[~sfSubject].value_or(account_);
+    auto const issuer = ctx_.tx[~sfIssuer].value_or(account_);
+
+    auto const credType(ctx_.tx[sfCredentialType]);
+    auto const sleCred =
+        view().peek(keylet::credential(subject, issuer, credType));
+
+    if ((subject != account_) && (issuer != account_))
+    {
+        if (checkNotExpired(sleCred, ctx_.view().info().parentCloseTime))
+        {
+            JLOG(j_.trace()) << "Can't delete non-expired credential.";
+            return tecNO_PERMISSION;
+        }
+    }
+
+    return deleteSLE(view(), sleCred, j_);
+}
+
+// ------- APPLY --------------------------
+
+NotTEC
+CredentialAccept::preflight(PreflightContext const& ctx)
+{
+    if (!ctx.rules.enabled(featureCredentials))
+    {
+        JLOG(ctx.j.trace()) << "featureCredentials is disabled.";
+        return temDISABLED;
+    }
+
+    if (auto const ret = preflight1(ctx); !isTesSuccess(ret))
+        return ret;
+
+    if (!ctx.tx[sfIssuer])
+    {
+        JLOG(ctx.j.trace()) << "Malformed transaction: Issuer field zeroed.";
+        return temINVALID_ACCOUNT_ID;
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+CredentialAccept::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const subject = ctx.tx[sfAccount];
+    AccountID const issuer = ctx.tx[sfIssuer];
+    auto const credType(ctx.tx[sfCredentialType]);
+
+    if (!ctx.view.exists(keylet::account(issuer)))
+    {
+        JLOG(ctx.j.warn()) << "No issuer: " << to_string(issuer);
+        return tecNO_ISSUER;
+    }
+
+    auto const sleCred =
+        ctx.view.read(keylet::credential(subject, issuer, credType));
+    if (!sleCred)
+    {
+        JLOG(ctx.j.warn()) << "No credential: " << to_string(subject) << ", "
+                           << to_string(issuer) << ", " << credType;
+        return tecNO_ENTRY;
+    }
+
+    if (sleCred->getFieldU32(sfFlags) & lsfAccepted)
+    {
+        JLOG(ctx.j.warn()) << "Credential already accepted: "
+                           << to_string(subject) << ", " << to_string(issuer)
+                           << ", " << credType;
+        return tecDUPLICATE;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+CredentialAccept::doApply()
+{
+    AccountID const subject{account_};
+    AccountID const issuer{ctx_.tx[sfIssuer]};
+
+    auto const sleSubj = view().peek(keylet::account(subject));
+    auto const sleIss = view().peek(keylet::account(issuer));
```

Comment:
Can you use the full name `sleIssuer` here, so it's less confusing?

Reply:
fixed